File: backend/app/schemas/b2b_portal.py
"""B2B 代理商门户 API 出入参模型。"""
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Annotated, Any

from pydantic import BaseModel, EmailStr, Field, PlainSerializer

# Decimal 序列化为 JSON number（而非 string），避免前端 toFixed 报错
DecimalFloat = Annotated[
    Decimal, PlainSerializer(lambda x: float(x), return_type=float, when_used="json")
]


# ============================================
# 认证
# ============================================

class B2BLoginRequest(BaseModel):
    email: EmailStr
    password: str


class B2BTokenResponse(BaseModel):
    access_token: str
    token_type: str = "bearer"
    agent: "B2BAgentProfile"


class B2BChangePasswordRequest(BaseModel):
    old_password: str
    new_password: str = Field(min_length=8, max_length=128)


# ============================================
# 代理商申请
# ============================================

class B2BApplicationRequest(BaseModel):
    """公开代理商申请表单。"""
    company_name: str = Field(min_length=2, max_length=200)
    contact_name: str = Field(min_length=2, max_length=100)
    email: EmailStr
    phone: str | None = Field(None, max_length=50)
    whatsapp: str | None = Field(None, max_length=50)
    wechat: str | None = Field(None, max_length=100)
    country: str | None = Field(None, max_length=64)
    city: str | None = Field(None, max_length=100)
    address: str | None = Field(None, max_length=500)
    password: str = Field(min_length=8, max_length=128)
    business_type: str | None = Field(None, max_length=100)
    website: str | None = Field(None, max_length=255)
    estimated_annual_volume: str | None = Field(None, max_length=100)
    product_interests: str | None = None
    message: str | None = None


class B2BApplicationResponse(BaseModel):
    id: str
    email: str
    company_name: str
    status: str
    message: str


# ============================================
# 代理商资料
# ============================================

class B2BAgentProfile(BaseModel):
    id: str
    agent_number: str
    company_name: str
    contact_name: str
    email: str
    phone: str | None = None
    country: str | None = None
    city: str | None = None
    address: str | None = None
    tier: str
    status: str
    commission_rate: DecimalFloat
    discount_percent: DecimalFloat
    credit_limit: DecimalFloat
    current_balance: DecimalFloat
    available_credit: DecimalFloat
    payment_terms_days: int
    currency: str
    last_login_at: datetime | None = None

    model_config = {"from_attributes": True}


class B2BAccountSummary(BaseModel):
    agent: B2BAgentProfile
    total_orders: int
    total_revenue: DecimalFloat
    pending_payments: DecimalFloat
    payment_due_date: date | None = None


# ============================================
# 商品
# ============================================

class B2BProductItem(BaseModel):
    id: str
    sku: str
    name: str
    description: str | None = None
    category: str | None = None
    brand: str | None = None
    wholesale_price: DecimalFloat
    retail_price: DecimalFloat | None = None
    moq: int
    currency: str
    in_stock: bool
    stock_quantity: int = 0
    images: list[str] = Field(default_factory=list)
    attributes: dict[str, Any] = Field(default_factory=dict)


class B2BProductListResponse(BaseModel):
    items: list[B2BProductItem]
    total: int
    page: int
    page_size: int


class B2BProductDetail(B2BProductItem):
    weight_kg: DecimalFloat | None = None
    dimensions: dict[str, Any] | None = None
    cost_note: str | None = None


# ============================================
# 购物车与下单
# ============================================

class B2BCartItem(BaseModel):
    product_id: str
    quantity: int = Field(ge=1)


class B2BCreateOrderRequest(BaseModel):
    items: list[B2BCartItem] = Field(min_length=1)
    shipping_address: dict[str, Any] | None = None
    notes: str | None = None


class B2BOrderItemResponse(BaseModel):
    id: str
    product_id: str
    product_name: str
    sku: str
    quantity: int
    unit_price: DecimalFloat
    subtotal: DecimalFloat

    model_config = {"from_attributes": True}


class B2BOrderResponse(BaseModel):
    id: str
    order_number: str
    status: str
    payment_status: str
    subtotal: DecimalFloat
    discount_amount: DecimalFloat
    shipping_cost: DecimalFloat
    total: DecimalFloat
    currency: str
    shipping_address: dict[str, Any]
    payment_due_date: date | None = None
    tracking_number: str | None = None
    tracking_carrier: str | None = None
    notes: str | None = None
    items: list[B2BOrderItemResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class B2BOrderListResponse(BaseModel):
    items: list[B2BOrderResponse]
    total: int
    page: int
    page_size: int


# ============================================
# 门户自助销售：RFQ、报价与合同
# ============================================

class B2BPortalRFQItemCreate(BaseModel):
    product_id: str
    quantity: int = Field(ge=1)
    target_unit_price: DecimalFloat | None = Field(default=None, gt=0)
    specifications: dict[str, Any] = Field(default_factory=dict)
    notes: str | None = Field(default=None, max_length=2000)


class B2BPortalRFQCreate(BaseModel):
    requested_currency: str = Field(default="USD", min_length=3, max_length=8)
    destination_country: str | None = Field(default=None, max_length=8)
    incoterm: str | None = Field(default=None, max_length=16)
    requested_delivery_date: date | None = None
    notes: str | None = Field(default=None, max_length=5000)
    items: list[B2BPortalRFQItemCreate] = Field(min_length=1)


class B2BPortalRFQItemResponse(BaseModel):
    id: str
    product_id: str
    sku_snapshot: str
    product_name_snapshot: str
    requested_quantity: int
    target_unit_price: DecimalFloat | None = None
    specifications: dict[str, Any] = Field(default_factory=dict)
    notes: str | None = None


class B2BPortalRFQResponse(BaseModel):
    id: str
    rfq_number: str
    status: str
    source: str
    requested_currency: str
    destination_country: str | None = None
    incoterm: str | None = None
    requested_delivery_date: date | None = None
    notes: str | None = None
    submitted_at: datetime | None = None
    closed_at: datetime | None = None
    items: list[B2BPortalRFQItemResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class B2BPortalRFQListResponse(BaseModel):
    items: list[B2BPortalRFQResponse]
    total: int
    page: int
    page_size: int


class B2BPortalQuoteItemResponse(BaseModel):
    id: str
    product_id: str
    sku_snapshot: str
    product_name_snapshot: str
    quantity: int
    unit_price: DecimalFloat
    discount_percent: DecimalFloat
    line_subtotal: DecimalFloat
    line_total: DecimalFloat
    price_source: str
    specifications: dict[str, Any] = Field(default_factory=dict)


class B2BPortalContractSummary(BaseModel):
    id: str
    contract_number: str
    status: str
    customer_signed_at: datetime | None = None
    company_signed_at: datetime | None = None
    activated_at: datetime | None = None


class B2BPortalQuoteResponse(BaseModel):
    id: str
    quote_number: str
    version_number: int
    rfq_id: str | None = None
    rfq_number: str | None = None
    status: str
    currency: str
    valid_until: date
    payment_terms_days: int
    incoterm: str | None = None
    shipping_terms: str | None = None
    subtotal: DecimalFloat
    discount_amount: DecimalFloat
    shipping_cost: DecimalFloat
    tax_amount: DecimalFloat
    total: DecimalFloat
    sent_at: datetime | None = None
    accepted_at: datetime | None = None
    rejected_at: datetime | None = None
    rejection_reason: str | None = None
    notes: str | None = None
    items: list[B2BPortalQuoteItemResponse] = Field(default_factory=list)
    contract: B2BPortalContractSummary | None = None
    created_at: datetime
    updated_at: datetime


class B2BPortalQuoteListResponse(BaseModel):
    items: list[B2BPortalQuoteResponse]
    total: int
    page: int
    page_size: int


class B2BPortalQuoteDecisionRequest(BaseModel):
    reason: str | None = Field(default=None, max_length=1000)


class B2BPortalContractResponse(BaseModel):
    id: str
    contract_number: str
    quote_id: str
    quote_number: str | None = None
    version_number: int | None = None
    status: str
    effective_from: date
    effective_to: date | None = None
    currency: str
    total: DecimalFloat
    document_url: str | None = None
    terms: dict[str, Any] = Field(default_factory=dict)
    customer_signed_by: str | None = None
    customer_signed_at: datetime | None = None
    company_signed_at: datetime | None = None
    activated_at: datetime | None = None
    terminated_at: datetime | None = None
    items: list[B2BPortalQuoteItemResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class B2BPortalContractListResponse(BaseModel):
    items: list[B2BPortalContractResponse]
    total: int
    page: int
    page_size: int


class B2BPortalContractSignRequest(BaseModel):
    signed_by: str = Field(min_length=1, max_length=128)


class B2BPortalOrderConversionResponse(BaseModel):
    id: str
    order_number: str
    quote_id: str
    contract_id: str
    status: str
    payment_status: str
    total: DecimalFloat
    currency: str


# 解决前向引用
B2BTokenResponse.model_rebuild()

File: frontend/src/api/b2bPortal.ts
export const B2B_PORTAL_TOKEN_KEY = 'b2b_portal_token'
export const B2B_PORTAL_BASE = '/api/v1/b2b-portal'

export class PortalApiError extends Error {
  status: number
  detail: unknown

  constructor(message: string, status: number, detail?: unknown) {
    super(message)
    this.name = 'PortalApiError'
    this.status = status
    this.detail = detail
  }
}

interface PortalRequestOptions extends RequestInit {
  timeoutMs?: number
}

export function getPortalToken(): string | null {
  return localStorage.getItem(B2B_PORTAL_TOKEN_KEY)
}

export function setPortalToken(token: string): void {
  localStorage.setItem(B2B_PORTAL_TOKEN_KEY, token)
}

export function clearPortalToken(): void {
  localStorage.removeItem(B2B_PORTAL_TOKEN_KEY)
}

function parseError(payload: unknown, fallback: string): string {
  if (typeof payload === 'string' && payload.trim()) return payload
  if (payload && typeof payload === 'object') {
    const record = payload as Record<string, unknown>
    if (typeof record.detail === 'string' && record.detail.trim()) {
      return record.detail
    }
    if (Array.isArray(record.detail) && record.detail.length > 0) {
      const first = record.detail[0] as Record<string, unknown>
      if (typeof first?.msg === 'string') return first.msg
    }
  }
  return fallback
}

export async function portalRequest<T>(
  endpoint: string,
  options: PortalRequestOptions = {},
): Promise<T> {
  const { timeoutMs = 30000, ...fetchOptions } = options
  const controller = new AbortController()
  const timer = window.setTimeout(() => controller.abort(), timeoutMs)
  const token = getPortalToken()
  const headers = new Headers(fetchOptions.headers)

  if (!(fetchOptions.body instanceof FormData) && !headers.has('Content-Type')) {
    headers.set('Content-Type', 'application/json')
  }
  if (token) {
    headers.set('Authorization', `Bearer ${token}`)
  }

  try {
    const response = await fetch(`${B2B_PORTAL_BASE}${endpoint}`, {
      ...fetchOptions,
      headers,
      signal: controller.signal,
    })
    const contentType = response.headers.get('content-type') || ''
    const payload = contentType.includes('application/json')
      ? await response.json().catch(() => null)
      : await response.text().catch(() => '')

    if (!response.ok) {
      if (response.status === 401) clearPortalToken()
      throw new PortalApiError(
        parseError(payload, `请求失败（HTTP ${response.status}）`),
        response.status,
        payload,
      )
    }
    return payload as T
  } catch (error) {
    if (error instanceof PortalApiError) throw error
    if (error instanceof DOMException && error.name === 'AbortError') {
      throw new PortalApiError('请求超时', 408)
    }
    throw new PortalApiError(error instanceof Error ? error.message : '网络请求失败', 0)
  } finally {
    window.clearTimeout(timer)
  }
}

export const portalApi = {
  login: (email: string, password: string) =>
    portalRequest<PortalLoginResponse>('/auth/login', {
      method: 'POST',
      body: JSON.stringify({ email, password }),
    }),
  apply: (data: Record<string, unknown>) =>
    portalRequest('/applications', {
      method: 'POST',
      body: JSON.stringify(data),
    }),
  me: () => portalRequest<PortalAgent>('/auth/me'),
  summary: () => portalRequest<PortalAccountSummary>('/account/summary'),
  products: (search?: string) => {
    const params = new URLSearchParams({ page: '1', page_size: '100' })
    if (search) params.set('search', search)
    return portalRequest<PortalProductList>(`/products?${params.toString()}`)
  },
  rfqs: (page = 1, pageSize = 100) =>
    portalRequest<PortalRFQList>(`/rfqs?page=${page}&page_size=${pageSize}`),
  createRfq: (data: Record<string, unknown>) =>
    portalRequest<PortalRFQ>('/rfqs', {
      method: 'POST',
      body: JSON.stringify(data),
    }),
  quotes: (page = 1, pageSize = 100) =>
    portalRequest<PortalQuoteList>(`/quotes?page=${page}&page_size=${pageSize}`),
  acceptQuote: (id: string) =>
    portalRequest<PortalQuote>(`/quotes/${id}/accept`, { method: 'POST' }),
  rejectQuote: (id: string, reason: string) =>
    portalRequest<PortalQuote>(`/quotes/${id}/reject`, {
      method: 'POST',
      body: JSON.stringify({ reason }),
    }),
  contracts: (page = 1, pageSize = 100) =>
    portalRequest<PortalContractList>(`/contracts?page=${page}&page_size=${pageSize}`),
  signContract: (id: string, signedBy: string) =>
    portalRequest<PortalContract>(`/contracts/${id}/sign`, {
      method: 'POST',
      body: JSON.stringify({ signed_by: signedBy }),
    }),
  convertContract: (id: string) =>
    portalRequest<PortalOrderConversion>(`/contracts/${id}/convert-to-order`, {
      method: 'POST',
    }),
  orders: (page = 1, pageSize = 100) =>
    portalRequest<PortalOrderList>(`/orders?page=${page}&page_size=${pageSize}`),
}

export interface PortalAgent {
  id: string
  agent_number: string
  company_name: string
  contact_name: string
  email: string
  phone: string | null
  country: string | null
  city: string | null
  address: string | null
  tier: string
  status: string
  commission_rate: number
  discount_percent: number
  credit_limit: number
  current_balance: number
  available_credit: number
  payment_terms_days: number
  currency: string
  last_login_at: string | null
}

export interface PortalLoginResponse {
  access_token: string
  token_type: string
  agent: PortalAgent
}

export interface PortalAccountSummary {
  agent: PortalAgent
  total_orders: number
  total_revenue: number
  pending_payments: number
  payment_due_date: string | null
}

export interface PortalProduct {
  id: string
  sku: string
  name: string
  description: string | null
  category: string | null
  brand: string | null
  wholesale_price: number
  retail_price: number | null
  moq: number
  currency: string
  in_stock: boolean
  stock_quantity: number
}

interface PortalProductList {
  items: PortalProduct[]
  total: number
}

export interface PortalRFQItem {
  id: string
  product_id: string
  sku_snapshot: string
  product_name_snapshot: string
  requested_quantity: number
  target_unit_price: number | null
  notes: string | null
}

export interface PortalRFQ {
  id: string
  rfq_number: string
  status: string
  source: string
  requested_currency: string
  destination_country: string | null
  incoterm: string | null
  requested_delivery_date: string | null
  notes: string | null
  submitted_at: string | null
  closed_at: string | null
  items: PortalRFQItem[]
  created_at: string
  updated_at: string
}

interface PortalRFQList {
  items: PortalRFQ[]
  total: number
}

export interface PortalQuoteItem {
  id: string
  product_id: string
  sku_snapshot: string
  product_name_snapshot: string
  quantity: number
  unit_price: number
  discount_percent: number
  line_subtotal: number
  line_total: number
  price_source: string
}

export interface PortalContractSummary {
  id: string
  contract_number: string
  status: string
  customer_signed_at: string | null
  company_signed_at: string | null
  activated_at: string | null
}

export interface PortalQuote {
  id: string
  quote_number: string
  version_number: number
  rfq_id: string | null
  rfq_number: string | null
  status: string
  currency: string
  valid_until: string
  payment_terms_days: number
  incoterm: string | null
  shipping_terms: string | null
  subtotal: number
  discount_amount: number
  shipping_cost: number
  tax_amount: number
  total: number
  sent_at: string | null
  accepted_at: string | null
  rejected_at: string | null
  rejection_reason: string | null
  notes: string | null
  items: PortalQuoteItem[]
  contract: PortalContractSummary | null
  created_at: string
  updated_at: string
}

interface PortalQuoteList {
  items: PortalQuote[]
  total: number
}

export interface PortalContract {
  id: string
  contract_number: string
  quote_id: string
  quote_number: string | null
  version_number: number | null
  status: string
  effective_from: string
  effective_to: string | null
  currency: string
  total: number
  document_url: string | null
  terms: Record<string, unknown>
  customer_signed_by: string | null
  customer_signed_at: string | null
  company_signed_at: string | null
  activated_at: string | null
  terminated_at: string | null
  items: PortalQuoteItem[]
  created_at: string
  updated_at: string
}

interface PortalContractList {
  items: PortalContract[]
  total: number
}

export interface PortalOrderItem {
  id: string
  product_id: string
  product_name: string
  sku: string
  quantity: number
  unit_price: number
  subtotal: number
}

export interface PortalOrder {
  id: string
  order_number: string
  status: string
  payment_status: string
  subtotal: number
  discount_amount: number
  shipping_cost: number
  total: number
  currency: string
  shipping_address: Record<string, unknown>
  payment_due_date: string | null
  tracking_number: string | null
  tracking_carrier: string | null
  notes: string | null
  items: PortalOrderItem[]
  created_at: string
  updated_at: string
}

interface PortalOrderList {
  items: PortalOrder[]
  total: number
}

export interface PortalOrderConversion {
  id: string
  order_number: string
  quote_id: string
  contract_id: string
  status: string
  payment_status: string
  total: number
  currency: string
}
